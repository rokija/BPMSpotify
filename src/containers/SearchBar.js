import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from "prop-types";
import { bindActionCreators } from "redux";
import { getSearchResults } from '../actions/searchActions';
import Cookies from 'universal-cookie';
const cookies = new Cookies();

export class SearchBar extends Component {
    constructor() {
        super();

        this.state = {
            term: '',
            errorMessage: ''
        };

        this.onInputChange = this.onInputChange.bind(this);
        this.onFormSubmit = this.onFormSubmit.bind(this);
    }

    componentWillMount() {
        const tokenData = cookies.get('token');

        if (tokenData) {
            return true;
        }
        else {
            this.context.router.history.push("/login");
        }
    }

    onInputChange(event) {
        this.setState({ term: event.target.value, errorMessage: '' });
    }

    onFormSubmit(event) {
        const {getSearchResults, setSearchQuery} = this.props;
        event.preventDefault();

        if (this.state.term !== '') {
            getSearchResults(this.state.term);
            setSearchQuery(this.state.term);
            this.setState({ term: '' });
        }
        else {
            this.setState({ errorMessage: 'please write a song or an artist name in the input' });
        }
    }

    render() {
        return (
            <div className="search-bar-wrapper">
                <form className="col-md-12 col-xs-12 search-input-wrapper">
                    <input
                        className="search-input"
                        placeholder="Search a song"
                        value={this.state.term}
                        onChange={this.onInputChange} />
                    <button
                        className="search-button btn glyphicon glyphicon-search"
                        onClick={this.onFormSubmit}
                        type="submit" />
                </form>
                <span className="col-md-12 col-xs-12 error-message-input">{this.state.errorMessage}</span>
            </div>
        );
    }
}


SearchBar.propTypes = {
    dispatch: PropTypes.func,
    searchTerm: PropTypes.string,
    getSearchResults: PropTypes.func,
    setSearchQuery: PropTypes.func,
};

SearchBar.contextTypes = {
    router: PropTypes.object
};


export function mapDispatchToProps(dispatch) {
    return bindActionCreators({ getSearchResults }, dispatch);
}

export default connect(null, mapDispatchToProps)(SearchBar);
