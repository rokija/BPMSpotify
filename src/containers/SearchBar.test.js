import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { SearchBar, mapDispatchToProps } from './SearchBar';
import Cookies from 'universal-cookie';

const cookies = new Cookies();

const context = { router: { history: [] }  };

const minProps ={
    getSearchResults: jest.fn(),
    setSearchQuery: jest.fn()
};

describe('SearchBar', () => {
    const wrapper = mount(<SearchBar {...minProps} />, { context });

    it('by calling method calls callback with provided value', () => {
        wrapper.instance().callback = jest.fn();
        wrapper.instance().method(3);
        expect(wrapper.instance().callback).toBeCalledWith(3);
    });

    it('renders properly', () => {
        expect(wrapper.length).toEqual(1);
    });

    it('calls typeof mapDispatchToProps', () => {
        expect(typeof (mapDispatchToProps())).toEqual("object");
    });

    it('renders properly if token is set', () => {
        cookies.set('token', 'dfsaasd', { path: '/' });
        const wrapper = shallow(<SearchBar />, { context });
        expect(wrapper.length).toEqual(1);
    });

    describe('simulating events', () => {
        it('calls onInputChange and add input value to the state', () => {
            wrapper.find('input').simulate('change', { target: { value: 'some value' } });
            expect(wrapper.state().term).toEqual('some value');
        });

        it('calls onFormSubmit and clears the state for term', () => {
            wrapper.find('button').simulate('click', {preventDefault: () => {}});
            expect(wrapper.state().term).toEqual('');
        });
    });

    describe('testing methods', () => {
        const wrapper = shallow(<SearchBar {...minProps} />, { context });
        const wrapperInstance = wrapper.instance();

        it('calls onInputChange', () => {
            wrapperInstance.onInputChange({ target: { value: 'some value' } });
            expect(wrapperInstance.state.term).toEqual('some value');
        });

        it('calls onFormSubmit', () => {
            wrapperInstance.setState({ term: 'value' });
            wrapperInstance.onFormSubmit({preventDefault: () => {}});
            expect(wrapperInstance.state.term).toEqual('');
        });

        it('calls onFormSubmit with no input value', () => {
            wrapperInstance.setState({ term: '' });
            wrapperInstance.onFormSubmit({preventDefault: () => {}});
            expect(wrapperInstance.state.term).toEqual('');
        });

        it('matches snapshot', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });
    });
});
