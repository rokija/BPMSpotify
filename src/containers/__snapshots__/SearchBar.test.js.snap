// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar testing methods matches snapshot 1`] = `
<SearchBar
  getSearchResults={[Function]}
  setSearchQuery={[Function]}
>
  <div
    className="search-bar-wrapper"
  >
    <form
      className="col-md-12 col-xs-12 search-input-wrapper"
    >
      <input
        className="search-input"
        onChange={[Function]}
        placeholder="Search a song"
        value=""
      />
      <button
        className="search-button btn glyphicon glyphicon-search"
        onClick={[Function]}
        type="submit"
      />
    </form>
    <span
      className="col-md-12 col-xs-12 error-message-input"
    />
    <Child>
      <div>
        <h2>
          CHILD
        </h2>
      </div>
    </Child>
  </div>
</SearchBar>
`;
